File: LeetCode/287_find_the_duplicate_number.rb
#Use the cycle detection algorithm because if we view the array like a function from 0..n to to 1..n it can't be injective (pigeon hole),
#so the linked list representation of the function will have a cycle.

# @param {Integer[]} nums
# @return {Integer}
def find_duplicate(nums)
    slow = 0
    fast = 0
    
    begin
        slow = nums[slow]
        fast = nums[nums[fast]]
    end while (slow != fast)
    
    find = 0
    
    while (find != slow)
        slow = nums[slow]
        find = nums[find]
    end
    
    find
end

File: LeetCode/287_find_the_duplicate_number.cpp
class Solution {
public:
    //One way to view this problem is finding the cycle if you follow the indices.
    //The start of the cycle will be the repeated element.
    int findDuplicate(vector<int>& nums) {
        size_t slow = 0;
        size_t fast = 0;
        
        do{
            slow = nums[slow];
            fast = nums[nums[fast]];
        } while (slow != fast);
        
        size_t finder = 0;
        
        while(slow != finder){
            slow = nums[slow];
            finder = nums[finder];
        }
        
        return finder;
    }
};